web/src/engine/osk/src/keyboard-layout/oskLayerGroup.ts

```diff
@@ -165,86 +171,97 @@ export default class OSKLayerGroup {
   }
 
   private nearestKey(coord: Omit<InputSample<KeyElement>, 'item'>, layer: OSKLayer): KeyElement {
-    const baseRect = this.element.getBoundingClientRect();
+    // If there are no rows, there are no keys; return instantly.
+    if(layer.rows.length == 0) {
+      return null;
+    }
+
+    // Our pre-processed layout info maps whatever shape the keyboard is in into a unit square.
+    // So, we map our coord to find its location within that square.
+    const proportionalCoords = {
+      // Note:  need to cache these two values in some manner; if the keyboard is swapped,
+      // we need them if any keypresses are registered before the swap completes.
+      x: coord.targetX / this.computedWidth,
+      y: coord.targetY / this.computedHeight
+    };
+
+    // If our computed width and/or height are 0, it's best to abort; key distance
+    // calculations are not viable.
+    if(!isFinite(proportionalCoords.x) || !isFinite(proportionalCoords.y)) {
+      return null;
+    }
 
+    // Step 1:  find the nearest row.
     let row: OSKRow = null;
     let bestMatchDistance = Number.MAX_VALUE;
 
+    // Max distance from the key's center to consider, vertically.
+    // Rows aren't variable-height - this value is "one size fits all."
+    const rowRadius = layer.rows[0].heightFraction / 2;
+
     // Find the row that the touch-coordinate lies within.
     for(const r of layer.rows) {
```

Comment:
Given rows are fixed height, presumably we shouldn't need a loop to do this calculation?

Reply:
I just double-checked how we calculate it to be safe:

https://github.com/keymanapp/keyman/blob/edadeb5729920c4e07fabc10ea83c48db9baee72/common/web/keyboard-processor/src/keyboards/activeLayout.ts#L700-L704

It does not appear that there's any tweaking of note for these values, so yes, they should map directly.  I was a bit worried that CSS styling may motivate minor tweaks that would make this slightly inaccurate at key edges.